database/db.go

```diff
@@ -0,0 +1,284 @@
+package database
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/SkynetLabs/skynet-accounts/build"
+	"github.com/SkynetLabs/skynet-accounts/database"
+	"github.com/sirupsen/logrus"
+	"gitlab.com/NebulousLabs/errors"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"go.mongodb.org/mongo-driver/mongo/readpref"
+	"go.sia.tech/siad/crypto"
+)
+
+var (
+	// CancelScanAfter defines how long we want to wait for a scan to finish
+	// before giving up on it and returning the skylink back into the "new"
+	// bucket, so the scan can be retried. This prevents scans from hanging
+	// forever in case the scanning server crashed or otherwise failed to
+	// either finish the scan or report its findings.
+	CancelScanAfter = build.Select(
+		build.Var{
+			Dev:      time.Minute,
+			Testing:  10 * time.Second,
+			Standard: time.Hour,
+		},
+	).(time.Duration)
+
+	// ErrNoDocumentsFound is returned when a database operation completes
+	// successfully but it doesn't find or affect any documents.
+	ErrNoDocumentsFound = errors.New("no documents found")
+	// ErrSkylinkExists is returned when we try to add a skylink to the database
+	// and it already exists there.
+	ErrSkylinkExists = errors.New("skylink already exists")
+
+	// True is a helper value, so we can pass a *bool to MongoDB's methods.
+	True = true
+
+	// dbName defines the name of the database this service uses
+	dbName = "scanner"
+	// dbSkylinks defines the name of the skylinks collection
+	dbSkylinks = "skylinks"
+	// mongoCompressors defines the compressors we are going to use for the
+	// connection to MongoDB
+	mongoCompressors = "zstd,zlib,snappy"
+	// mongoReadPreference defines the DB's read preference. The options are:
+	// primary, primaryPreferred, secondary, secondaryPreferred, nearest.
+	// See https://docs.mongodb.com/manual/core/read-preference/
+	mongoReadPreference = "nearest"
+	// mongoWriteConcern describes the level of acknowledgment requested from
+	// MongoDB.
+	mongoWriteConcern = "majority"
+	// mongoWriteConcernTimeout specifies a time limit, in milliseconds, for
+	// the write concern to be satisfied.
+	mongoWriteConcernTimeout = "1000"
+)
+
+// DB holds a connection to the database, as well as helpful shortcuts to
+// collections and utilities.
+type DB struct {
+	DB       *mongo.Database
+	Skylinks *mongo.Collection
+	Logger   *logrus.Logger
+}
+
+// New creates a new database connection.
+func New(ctx context.Context, creds database.DBCredentials, logger *logrus.Logger) (*DB, error) {
+	if logger == nil {
+		logger = &logrus.Logger{}
+	}
+	connStr := fmt.Sprintf(
+		"mongodb://%s:%s@%s:%s/?compressors=%s&readPreference=%s&w=%s&wtimeoutMS=%s",
```

Comment:
Is there a reason you used the client options in `skyd` instead of using a string as we do in `accounts`? :P

The answer is no, I just used whatever approach I found in the guides I read. If you think there is an advantage to using the client options, we can use that. Personally, I find the string approach way more readable, so I would prefer to continue using it.

Reply:
haha yeah. We usually choose strong types, constants and methods over hardcoded values. It just leaves less room for errors and allows for static analysis and deprecation as part of the library. In some cases setters also touch multiple things or do some input validation which can be helpful.

Reply:
Done.